src/librustc_data_structures/profiling.rs

```diff
@@ -585,3 +601,82 @@ fn get_resident() -> Option<usize> {
         }
     }
 }
+
+fn pass_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| profiler.alloc_string(llvm_name))
+}
+
+fn ir_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| {
+        let demangled_ir_name = rustc_demangle::demangle(llvm_name).to_string();
+        profiler.alloc_string(demangled_ir_name.as_str())
+    })
+}
+
+fn llvm_args_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    pass_name: &str,
+    ir_name: &str,
+) -> EventId {
+    let pass_name = pass_name_to_string_id(profiler, string_cache, pass_name);
+    let mut components = vec![StringComponent::Ref(pass_name)];
+    // handle that LazyCallGraph::SCC is a comma separated list within parentheses
+    let parentheses: &[_] = &['(', ')'];
+    let trimed = ir_name.trim_matches(parentheses);
+    for part in trimed.split(',') {
+        let ir_name = ir_name_to_string_id(profiler, string_cache, part);
+        components.push(StringComponent::Value(SEPARATOR_BYTE));
+        components.push(StringComponent::Ref(ir_name));
+    }
```

Comment:
Is there ever more than one argument? If not you can use the `EventIdBuilder::from_label_and_arg`.

Reply:
for LazyCallGraph::SCC there can be up to 10 args (if more they are truncated by llvm) do not know if it is interesting to see all the functions in the SCC (and in that case make a new getName function in llvm that returns all not only 10) or if it is enough to print the first function.  see around https://github.com/rust-lang/llvm-project/blob/a6f4c1bb07d58df5956d2c49be68547143f77b18/llvm/include/llvm/Analysis/LazyCallGraph.h#L447 for more info.

Reply:
Yeah we'll want to add a `EventIdBuilder::from_label_and_args(label: _, args: &[_])` method at some point anyway. For now encoding things directly is fine.